cmd/peer-rest-client.go

```diff
@@ -0,0 +1,418 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/gob"
+	"io"
+	"net/url"
+
+	"github.com/minio/minio/cmd/http"
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/cmd/rest"
+	"github.com/minio/minio/pkg/event"
+	xnet "github.com/minio/minio/pkg/net"
+	"github.com/minio/minio/pkg/policy"
+)
+
+// Abstracts a remote disk.
+type peerRESTClient struct {
+	host       *xnet.Host
+	restClient *rest.Client
+	connected  bool
+	lastError  error
+	instanceID string // REST server's instanceID which is sent with every request for validation.
+}
+
+// Wrapper to restClient.Call to handle network errors, in case of network error the connection is marked disconnected
+// permanently. The only way to restore the connection is at the xl-sets layer by xlsets.monitorAndConnectEndpoints()
+// after verifying format.json
+func (client *peerRESTClient) call(method string, values url.Values, body io.Reader, length int64) (respBody io.ReadCloser, err error) {
+	if !client.connected {
+		return nil, errNodeConnection
+	}
+
+	if values == nil {
+		values = make(url.Values)
+	}
+	values.Set(peerRESTInstanceID, client.instanceID)
+
+	respBody, err = client.restClient.Call(method, values, body, length)
+	if err == nil {
+		return respBody, nil
+	}
+	client.lastError = err
+	if isNetworkError(err) {
+		client.connected = false
+	}
+
+	return nil, err
+}
+
+// Gets peer server's instanceID - to be used with every REST call for validation.
+func (client *peerRESTClient) getInstanceID() (err error) {
```

Comment:
`getInstanceID()` is needed only for storage-rest calls to detect reboots and revalidate format.json so it can be removed for peer-to-peer notification type calls

Reply:
done